From ffb40f05e5babf8c725c93914c5acff8758c9673 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Tue, 14 Mar 2017 10:45:10 +0100
Subject: [PATCH] TOOLS-1717 sdcadm should not load SDC Application from SAPI
 when not required Reviewed by: Trent Mick <trentm@gmail.com>

---
 lib/channel.js                        | 208 +++++++++-------
 lib/cli/do_avail.js                   |  17 +-
 lib/cli/do_update_gz_tools.js         |  34 +--
 lib/cli/do_update_other.js            |  19 +-
 lib/cli/index.js                      |   1 -
 lib/dc-maint.js                       |   2 +-
 lib/default-fabric.js                 |   7 +-
 lib/platform.js                       |  87 ++++---
 lib/post-setup/cmon.js                |  15 +-
 lib/post-setup/cns.js                 |  11 +-
 lib/post-setup/dev-sample-data.js     |  43 ++--
 lib/post-setup/docker.js              |   7 +-
 lib/post-setup/fabrics.js             | 179 ++++++++------
 lib/post-setup/ha-binder.js           |   8 +-
 lib/post-setup/ha-manatee.js          |   6 +-
 lib/post-setup/underlay-nics.js       |   8 +-
 lib/procedures/update-agent-v1.js     |   5 +-
 lib/procedures/update-dockerlogger.js |   8 +-
 lib/sdcadm.js                         | 340 ++++++++++++++++----------
 lib/steps/agent-services.js           |   3 +-
 lib/steps/no-rabbit.js                |   9 +-
 21 files changed, 622 insertions(+), 395 deletions(-)

diff --git a/lib/channel.js b/lib/channel.js
index 23031ab..62a1b2e 100644
--- a/lib/channel.js
+++ b/lib/channel.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -18,7 +18,7 @@
 
 var util = require('util');
 var tabula = require('tabula');
-
+var vasync = require('vasync');
 var cmdln = require('cmdln');
 var Cmdln = cmdln.Cmdln;
 
@@ -59,63 +59,76 @@ function do_list(subcmd, opts, args, cb) {
     }
 
     var progress = self.progress;
-    var app = self.sdcadm.sdc;
+    var app;
 
-    self.sdcadm.updates.listChannels({}, function (err, channels) {
-        if (err) {
-            progress('Error trying to retrieve update channels');
-            var e = new errors.SDCClientError(err, 'imgapi');
-            return cb(e);
-        }
+    vasync.pipeline({arg: {}, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
+        function getChannels(ctx, next) {
+            app = self.sdcadm.sdcApp;
+            self.sdcadm.updates.listChannels({}, function (err, channels) {
+                if (err) {
+                    progress('Error trying to retrieve update channels');
+                    var e = new errors.SDCClientError(err, 'imgapi');
+                    next(e);
+                    return;
+                }
 
-        if (app.metadata.update_channel) {
-            channels = channels.map(function (c) {
-                if (c.name === app.metadata.update_channel) {
-                    c['default'] = true;
+                if (app.metadata.update_channel) {
+                    ctx.channels = channels.map(function (c) {
+                        if (c.name === app.metadata.update_channel) {
+                            c['default'] = true;
+                        } else {
+                            delete c['default'];
+                        }
+                        return c;
+                    });
                 } else {
-                    delete c['default'];
+                    ctx.channels = channels.map(function (c) {
+                        if (c['default']) {
+                            c.remote = true;
+                        }
+                        return c;
+                    });
                 }
-                return c;
+
+                next();
             });
-        } else {
-            channels = channels.map(function (c) {
-                if (c['default']) {
-                    c.remote = true;
+        },
+        function printChannels(ctx, next) {
+            if (opts.json) {
+                console.log(JSON.stringify(ctx.channels, null, 4));
+                next();
+                return;
+            }
+
+            ctx.channels = ctx.channels.map(function (c) {
+                if (c['default'] && c.remote) {
+                    delete c.remote;
+                    c['default'] = 'true (remote)';
                 }
                 return c;
             });
-        }
-
-        if (opts.json) {
-            console.log(JSON.stringify(channels, null, 4));
-            return cb();
-        }
-
-        channels = channels.map(function (c) {
-            if (c['default'] && c.remote) {
-                delete c.remote;
-                c['default'] = 'true (remote)';
-            }
-            return c;
-        });
 
 
-        var validFieldsMap = {};
-        channels.forEach(function (v) {
-            var k;
-            for (k in v) {
-                validFieldsMap[k] = true;
-            }
-        });
-
-        tabula(channels, {
-            skipHeader: opts.H,
-            columns: ['name', 'default', 'description'],
-            validFields: Object.keys(validFieldsMap)
-        });
-
-        return cb();
+            var validFieldsMap = {};
+            ctx.channels.forEach(function (v) {
+                var k;
+                for (k in v) {
+                    validFieldsMap[k] = true;
+                }
+            });
 
+            tabula(ctx.channels, {
+                skipHeader: opts.H,
+                columns: ['name', 'default', 'description'],
+                validFields: Object.keys(validFieldsMap)
+            });
+            next();
+        }
+    ]}, function pipeCb(pipeErr) {
+        cb(pipeErr);
     });
 };
 
@@ -162,36 +175,52 @@ function do_set(subcmd, opts, args, cb) {
 
     var channel = args.shift();
     var progress = self.progress;
-    var app = self.sdcadm.sdc;
-
-    self.sdcadm.updates.listChannels({}, function (err, channels) {
-        if (err) {
-            progress('Error trying to retrieve update channels');
-            var e = new errors.SDCClientError(err, 'imgapi');
-            return cb(e);
-        }
 
-        var names = channels.map(function (c) {
-            return (c.name);
-        });
-
-        if (names.indexOf(channel) === -1) {
-            progress('Must specify a valid channel: %j', channels);
-            return cb(new errors.UsageError('Invalid channel name'));
-        }
+    vasync.pipeline({arg: {}, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
+        function getChannels(ctx, next) {
+            self.sdcadm.updates.listChannels({}, function (err, channels) {
+                if (err) {
+                    progress('Error trying to retrieve update channels');
+                    var e = new errors.SDCClientError(err, 'imgapi');
+                    next(e);
+                    return;
+                }
+                ctx.channels = channels;
+                next();
+            });
+        },
+        function validateChannel(ctx, next) {
+            var names = ctx.channels.map(function (c) {
+                return (c.name);
+            });
 
-        self.sdcadm.sapi.updateApplication(app.uuid, {
-            metadata: {
-                update_channel: channel
-            }
-        }, function (updateErr, svc) {
-            if (updateErr) {
-                return cb(new errors.SDCClientError(updateErr, 'sapi'));
+            if (names.indexOf(channel) === -1) {
+                progress('Must specify a valid channel: %j', ctx.channels);
+                next(new errors.UsageError('Invalid channel name'));
+                return;
             }
-            progress('Update channel has been successfully set to: \'%s\'',
-                    channel);
-            return cb();
-        });
+            next();
+        },
+        function setChannel(ctx, next) {
+            self.sdcadm.sapi.updateApplication(self.sdcadm.sdcApp.uuid, {
+                metadata: {
+                    update_channel: channel
+                }
+            }, function (updateErr, svc) {
+                if (updateErr) {
+                    next(new errors.SDCClientError(updateErr, 'sapi'));
+                    return;
+                }
+                progress('Update channel has been successfully set to: \'%s\'',
+                        channel);
+                next();
+            });
+        }
+    ]}, function pipeCb(pipeErr) {
+        cb(pipeErr);
     });
 };
 
@@ -222,19 +251,28 @@ function do_unset(subcmd, opts, args, cb) {
     }
 
     var progress = self.progress;
-    var app = self.sdcadm.sdc;
 
-    self.sdcadm.sapi.updateApplication(app.uuid, {
-        metadata: {
-            update_channel: app.metadata.update_channel
+    vasync.pipeline({arg: {}, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
         },
-        action: 'delete'
-    }, function (err, svc) {
-        if (err) {
-            return cb(new errors.SDCClientError(err, 'sapi'));
+        function unsetChannel(_, next) {
+            self.sdcadm.sapi.updateApplication(self.sdcadm.sdcApp.uuid, {
+                metadata: {
+                    update_channel: self.sdcadm.sdcApp.metadata.update_channel
+                },
+                action: 'delete'
+            }, function (err, svc) {
+                if (err) {
+                    next(new errors.SDCClientError(err, 'sapi'));
+                    return;
+                }
+                progress('Update channel has been successfully unset');
+                next();
+            });
         }
-        progress('Update channel has been successfully unset');
-        return cb();
+    ]}, function pipeCb(pipeErr) {
+        cb(pipeErr);
     });
 };
 
diff --git a/lib/cli/do_avail.js b/lib/cli/do_avail.js
index cf6a528..b7017bb 100644
--- a/lib/cli/do_avail.js
+++ b/lib/cli/do_avail.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -55,16 +55,23 @@ Available.prototype.execute = function cExecute(opts, args, cb) {
     var changes;
     var plan;
 
-    // Set or override the default channel if anything is given:
-    if (opts.channel) {
-        self.sdcadm.updates.channel = opts.channel;
-    }
+
 
     // override to true list portolan/rabbit
     opts.force_data_path = true;
     opts.force_rabbitmq = true;
 
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
+        function setChannel(_, next) {
+            // Set or override the default channel if anything is given:
+            if (opts.channel) {
+                self.sdcadm.updates.channel = opts.channel;
+            }
+            next();
+        },
         function getChangesFromArgs(_, next) {
             self.cli._specFromArgs(opts, args, function (err, chgs) {
                 if (err) {
diff --git a/lib/cli/do_update_gz_tools.js b/lib/cli/do_update_gz_tools.js
index 1377d5a..2293d21 100644
--- a/lib/cli/do_update_gz_tools.js
+++ b/lib/cli/do_update_gz_tools.js
@@ -5,10 +5,11 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var assert = require('assert-plus');
+var vasync = require('vasync');
 
 var errors = require('../errors');
 
@@ -48,19 +49,24 @@ function do_update_gz_tools(subcmd, opts, args, cb) {
             'must specify installer image UUID or --latest'));
     }
 
-    // Set or override the default channel if anything is given:
-    if (opts.channel) {
-        self.sdcadm.updates.channel = opts.channel;
-    }
-
-    self.sdcadm.updateGzTools({
-        image: opts.latest ? 'latest' : args[0],
-        progress: progress,
-        justDownload: opts.just_download,
-        forceReinstall: opts.force_reinstall,
-        concurrency: opts.concurrency
-    }, finish);
-
+    vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
+        function updateGzTools(_, next) {
+            // Set or override the default channel if anything is given:
+            if (opts.channel) {
+                self.sdcadm.updates.channel = opts.channel;
+            }
+            self.sdcadm.updateGzTools({
+                image: opts.latest ? 'latest' : args[0],
+                progress: progress,
+                justDownload: opts.just_download,
+                forceReinstall: opts.force_reinstall,
+                concurrency: opts.concurrency
+            }, next);
+        }
+    ]}, finish);
 }
 do_update_gz_tools.options = [
     {
diff --git a/lib/cli/do_update_other.js b/lib/cli/do_update_other.js
index 7e33b65..48b41c4 100644
--- a/lib/cli/do_update_other.js
+++ b/lib/cli/do_update_other.js
@@ -67,7 +67,7 @@ function do_update_other(subcmd, opts, args, cb) {
     }
 
     function updateSdcApp(svcOpts, next) {
-        var uuid = self.sdcadm.sdc.uuid;
+        var uuid = self.sdcadm.sdcApp.uuid;
         self.sdcadm.sapi.updateApplication(uuid, svcOpts, function (err, svc) {
             if (err) {
                 return next(new errors.SDCClientError(err, 'sapi'));
@@ -122,6 +122,9 @@ function do_update_other(subcmd, opts, args, cb) {
         progress: self.progress
     };
     vasync.pipeline({arg: context, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         /*
          * Time to finally make the switch to the new agents by default.
          */
@@ -171,11 +174,11 @@ function do_update_other(subcmd, opts, args, cb) {
 
         // Remove deprecated params.resolvers:
         function removeSdcAppResolvers(ctx, next) {
-            if (!self.sdcadm.sdc.params.resolvers) {
+            if (!self.sdcadm.sdcApp.params.resolvers) {
                 return next();
             }
             progress('Remove deprecated "sdc" SAPI app params resolvers');
-            self.sdcadm.sapi.updateApplication(self.sdcadm.sdc.uuid, {
+            self.sdcadm.sapi.updateApplication(self.sdcadm.sdcApp.uuid, {
                 action: 'delete',
                 params: {
                     resolvers: []
@@ -191,7 +194,7 @@ function do_update_other(subcmd, opts, args, cb) {
         },
 
         function updateSdcAppSchemas(ctx, next) {
-            var currSchema = self.sdcadm.sdc.metadata_schema;
+            var currSchema = self.sdcadm.sdcApp.metadata_schema;
             var latestSchema = schemas.sdc.sdc_app;
             if (currSchema && jsprim.deepEqual(currSchema, latestSchema)) {
                 return next();
@@ -200,7 +203,7 @@ function do_update_other(subcmd, opts, args, cb) {
             self.log.debug({before: currSchema, after: latestSchema},
                 'update sdc app metadata_schema');
             progress('Update "sdc" SAPI app metadata_schema');
-            self.sdcadm.sapi.updateApplication(self.sdcadm.sdc.uuid, {
+            self.sdcadm.sapi.updateApplication(self.sdcadm.sdcApp.uuid, {
                 action: 'update',
                 metadata_schema: latestSchema
             }, function (err, app) {
@@ -245,7 +248,7 @@ function do_update_other(subcmd, opts, args, cb) {
             vasync.forEachParallel({
                 inputs: svcsToUpdate,
                 func: function updateSvc(svc, nextSvc) {
-                    var mdata = self.sdcadm.sdc.metadata;
+                    var mdata = self.sdcadm.sdcApp.metadata;
                     var svcDomain = format('%s.%s.%s', svc.name,
                         mdata.datacenter_name, mdata.dns_domain);
                     progress('Set "%s" service "metadata.SERVICE_DOMAIN"',
@@ -263,7 +266,7 @@ function do_update_other(subcmd, opts, args, cb) {
         },
 
         function updateAppDomains(ctx, next) {
-            var mdata = self.sdcadm.sdc.metadata;
+            var mdata = self.sdcadm.sdcApp.metadata;
             var mdataUpdates = {};
 
             NEW_SERVICES.forEach(function (svcName) {
@@ -591,7 +594,7 @@ function do_update_other(subcmd, opts, args, cb) {
                 }
 
                 progress('Appending "sapi_domain" to node.config');
-                var mdata = self.sdcadm.sdc.metadata;
+                var mdata = self.sdcadm.sdcApp.metadata;
                 var sapiDomain = format('sapi_domain=\'sapi.%s.%s\'\n',
                     mdata.datacenter_name, mdata.dns_domain);
                 fs.appendFile(nodeConfig, sapiDomain, next);
diff --git a/lib/cli/index.js b/lib/cli/index.js
index e887624..9dd52e7 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -182,7 +182,6 @@ CLI.prototype.init = function init(opts, args, callback) {
             callback();
             return;
         }
-
         self.sdcadm = new SdcAdm({
             log: self.log,
             uuid: self.uuid
diff --git a/lib/dc-maint.js b/lib/dc-maint.js
index 28125c2..c04e071 100644
--- a/lib/dc-maint.js
+++ b/lib/dc-maint.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
diff --git a/lib/default-fabric.js b/lib/default-fabric.js
index 97027e0..5060a4f 100644
--- a/lib/default-fabric.js
+++ b/lib/default-fabric.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -63,7 +63,7 @@ function defFabricAddNetwork(opts, cb) {
        provision_start_ip: '192.168.128.5',
        provision_end_ip: '192.168.131.250',
        gateway: '192.168.128.1',
-       resolvers: opts.sdcadm.sdc.metadata.dns_resolvers.split(','),
+       resolvers: opts.sdcadm.sdcApp.metadata.dns_resolvers.split(','),
        vlan_id: 2
     };
 
@@ -109,6 +109,9 @@ function addDefaultFabric(opts, cb) {
     assert.func(cb, 'cb');
 
     vasync.pipeline({ arg: opts, funcs: [
+        function ensureSdcApp(_, next) {
+            opts.sdcadm.ensureSdcApp({}, next);
+        },
         defFabricAddVLAN,
         defFabricAddNetwork
     ]}, function (err, results) {
diff --git a/lib/platform.js b/lib/platform.js
index dbbb9a4..4bfeaa2 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -346,38 +346,55 @@ Platform.prototype.available = function available(opts, cb) {
     assert.optionalString(opts.channel, 'opts.channel');
 
     var self = this;
-
-    // Set or override the default channel if anything is given:
-    if (opts.channel) {
-        self.sdcadm.updates.channel = opts.channel;
-    }
-
-    self.getLatestPlatformInstalled(function (err2, latest) {
-        if (err2) {
-            return cb(err2);
-        }
-        var filter = {
-            name: 'platform'
-        };
-        self.sdcadm.updates.listImages(filter, function (err, images) {
-            if (err) {
-                return cb(new errors.SDCClientError(err, 'updates'));
-            }
-            if (Array.isArray(images) && !images.length) {
-                return cb(new errors.UpdateError('no images found'));
+    var imgs;
+    vasync.pipeline({arg: {}, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
+        function getLatestInstalled(ctx, next) {
+            // Set or override the default channel if anything is given:
+            if (opts.channel) {
+                self.sdcadm.updates.channel = opts.channel;
             }
-            common.sortArrayOfObjects(images, ['published_at']);
-            images = images.map(function (img) {
-                return ({
-                    version: img.version.split('-').pop(),
-                    uuid: img.uuid,
-                    published_at: img.published_at
+            self.getLatestPlatformInstalled(function (err2, latest) {
+                if (err2) {
+                    next(err2);
+                    return;
+                }
+                ctx.latest = latest;
+                next();
+            });
+        },
+        function getAvailableImages(ctx, next) {
+            var filter = {
+                name: 'platform'
+            };
+            self.sdcadm.updates.listImages(filter, function (err, images) {
+                if (err) {
+                    next(new errors.SDCClientError(err, 'updates'));
+                    return;
+                }
+                if (Array.isArray(images) && !images.length) {
+                    next(new errors.UpdateError('no images found'));
+                    return;
+                }
+                common.sortArrayOfObjects(images, ['published_at']);
+                images = images.map(function (img) {
+                    return ({
+                        version: img.version.split('-').pop(),
+                        uuid: img.uuid,
+                        published_at: img.published_at
+                    });
+                }).filter(function (i) {
+                    return (i.version > ctx.latest);
                 });
-            }).filter(function (i) {
-                return (i.version > latest);
+                imgs = images;
+                next(null);
             });
-            return cb(null, images);
-        });
+
+        }
+    ]}, function pipeCb(pipeErr) {
+        cb(pipeErr, imgs);
     });
 };
 
@@ -443,11 +460,6 @@ Platform.prototype.install = function install(opts, callback) {
     // image in order to avoid same thing than for TOOLS-876
     var imgNotFoundError = false;
 
-    // Set or override the default channel if anything is given:
-    if (opts.channel) {
-        self.sdcadm.updates.channel = opts.channel;
-    }
-
     function findPlatformImageLatest(cb) {
         var filter = {
             name: 'platform'
@@ -583,7 +595,14 @@ Platform.prototype.install = function install(opts, callback) {
     }
 
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         function getChannel(_, next) {
+            // Set or override the default channel if anything is given:
+            if (opts.channel) {
+                self.sdcadm.updates.channel = opts.channel;
+            }
             self.sdcadm.getDefaultChannel(function (err, channel) {
                 // Will not fail the whole operation due to channel not found
                 if (err) {
diff --git a/lib/post-setup/cmon.js b/lib/post-setup/cmon.js
index 5e90117..3611b74 100644
--- a/lib/post-setup/cmon.js
+++ b/lib/post-setup/cmon.js
@@ -78,6 +78,9 @@ function do_cmon(subcmd, opts, args, cb) {
     };
 
     vasync.pipeline({arg: context, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         function ensureSapiMode(_, next) {
             self.sdcadm.sapi.getMode(function (err, mode) {
                 if (err) {
@@ -94,7 +97,7 @@ function do_cmon(subcmd, opts, args, cb) {
         function ensureCnsSvc(ctx, next) {
             self.sdcadm.sapi.listServices({
                 name: 'cns',
-                application_uuid: self.sdcadm.sdc.uuid
+                application_uuid: self.sdcadm.sdcApp.uuid
             }, function (svcErr, svcs) {
                 if (svcErr) {
                     next(svcErr);
@@ -130,7 +133,7 @@ function do_cmon(subcmd, opts, args, cb) {
         function getSvc(ctx, next) {
             self.sdcadm.sapi.listServices({
                 name: 'cmon',
-                application_uuid: self.sdcadm.sdc.uuid
+                application_uuid: self.sdcadm.sdcApp.uuid
             }, function (svcErr, svcs) {
                 if (svcErr) {
                     return next(svcErr);
@@ -263,8 +266,8 @@ function do_cmon(subcmd, opts, args, cb) {
                 return;
             }
 
-            var domain = self.sdcadm.sdc.metadata.datacenter_name + '.' +
-                    self.sdcadm.sdc.metadata.dns_domain;
+            var domain = self.sdcadm.sdcApp.metadata.datacenter_name + '.' +
+                    self.sdcadm.sdcApp.metadata.dns_domain;
             var svcDomain = svcData.name + '.' + domain;
 
             self.progress('Creating "cmon" service');
@@ -275,7 +278,7 @@ function do_cmon(subcmd, opts, args, cb) {
             svcData.params.billing_id = ctx.cmonPkg.uuid;
             delete svcData.params.package_name;
 
-            self.sdcadm.sapi.createService('cmon', self.sdcadm.sdc.uuid,
+            self.sdcadm.sapi.createService('cmon', self.sdcadm.sdcApp.uuid,
                     svcData, function (err, svc) {
                 if (err) {
                     return next(new errors.SDCClientError(err, 'sapi'));
@@ -338,7 +341,7 @@ function do_cmon(subcmd, opts, args, cb) {
                 }
                 self.progress('Creating "cmon-agent" service');
                 self.sdcadm.sapi.createService('cmon-agent',
-                        self.sdcadm.sdc.uuid, {
+                        self.sdcadm.sdcApp.uuid, {
                     params: {
                         image_uuid: ctx.cmonAgentImg.uuid,
                         tags: {
diff --git a/lib/post-setup/cns.js b/lib/post-setup/cns.js
index a9eff6d..e4be048 100644
--- a/lib/post-setup/cns.js
+++ b/lib/post-setup/cns.js
@@ -67,6 +67,9 @@ function do_cns(subcmd, opts, args, cb) {
         didSomething: false
     };
     vasync.pipeline({arg: context, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         /* @field ctx.cnsPkg */
         function getPkg(ctx, next) {
             var filter = {name: svcData.params.package_name,
@@ -102,7 +105,7 @@ function do_cns(subcmd, opts, args, cb) {
         function getSvc(ctx, next) {
             self.sdcadm.sapi.listServices({
                 name: 'cns',
-                application_uuid: self.sdcadm.sdc.uuid
+                application_uuid: self.sdcadm.sdcApp.uuid
             }, function (svcErr, svcs) {
                 if (svcErr) {
                     return next(svcErr);
@@ -191,8 +194,8 @@ function do_cns(subcmd, opts, args, cb) {
                 return next();
             }
 
-            var domain = self.sdcadm.sdc.metadata.datacenter_name + '.' +
-                    self.sdcadm.sdc.metadata.dns_domain;
+            var domain = self.sdcadm.sdcApp.metadata.datacenter_name + '.' +
+                    self.sdcadm.sdcApp.metadata.dns_domain;
             var svcDomain = svcData.name + '.' + domain;
 
             self.progress('Creating "cns" service');
@@ -203,7 +206,7 @@ function do_cns(subcmd, opts, args, cb) {
             svcData.params.billing_id = ctx.cnsPkg.uuid;
             delete svcData.params.package_name;
 
-            self.sdcadm.sapi.createService('cns', self.sdcadm.sdc.uuid,
+            self.sdcadm.sapi.createService('cns', self.sdcadm.sdcApp.uuid,
                     svcData, function (err, svc) {
                 if (err) {
                     return next(new errors.SDCClientError(err, 'sapi'));
diff --git a/lib/post-setup/dev-sample-data.js b/lib/post-setup/dev-sample-data.js
index aeda52b..e6ac223 100644
--- a/lib/post-setup/dev-sample-data.js
+++ b/lib/post-setup/dev-sample-data.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -25,23 +25,6 @@ var errors = require('../errors');
 
 //---- internal support stuff
 
-function addDevSampleData(opts, cb) {
-    assert.object(opts, 'opts');
-    assert.object(opts.sdcadm, 'opts.sdcadm');
-    assert.object(opts.log, 'opts.log');
-    assert.func(opts.progress, 'opts.progress');
-    assert.func(cb, 'cb');
-
-    vasync.parallel({funcs: [
-        function pkgs(next) {
-            addDevSamplePkgs(opts, next);
-        },
-        function imgs(next) {
-            addDevSampleImgs(opts, next);
-        }
-    ]}, cb);
-}
-
 function addDevSamplePkgs(opts, cb) {
     var papi = opts.sdcadm.papi;
 
@@ -140,6 +123,30 @@ function addDevSampleImgs(opts, cb) {
 }
 
 
+function addDevSampleData(opts, cb) {
+    assert.object(opts, 'opts');
+    assert.object(opts.sdcadm, 'opts.sdcadm');
+    assert.object(opts.log, 'opts.log');
+    assert.func(opts.progress, 'opts.progress');
+    assert.func(cb, 'cb');
+
+    vasync.pipeline({funcs: [
+        function ensureSdcApp(_, nextFun) {
+            opts.sdcadm.ensureSdcApp({}, nextFun);
+        },
+        function addSampleData(_, nextFun) {
+            vasync.parallel({funcs: [
+                function pkgs(next) {
+                    addDevSamplePkgs(opts, next);
+                },
+                function imgs(next) {
+                    addDevSampleImgs(opts, next);
+                }
+            ]}, nextFun);
+        }
+    ]}, cb);
+}
+
 //---- CLI
 
 function do_dev_sample_data(subcmd, opts, args, cb) {
diff --git a/lib/post-setup/docker.js b/lib/post-setup/docker.js
index 720b270..8f16169 100644
--- a/lib/post-setup/docker.js
+++ b/lib/post-setup/docker.js
@@ -77,6 +77,9 @@ function do_docker(subcmd, opts, args, cb) {
         urConnection: null
     };
     vasync.pipeline({arg: context, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         /* @field ctx.dockerPkg */
         function getDockerPkg(ctx, next) {
             var filter = {name: dockerSvcData.params.package_name,
@@ -110,7 +113,7 @@ function do_docker(subcmd, opts, args, cb) {
         },
 
         function getSdcApp(ctx, next) {
-            ctx.app = self.sdcadm.sdc;
+            ctx.app = self.sdcadm.sdcApp;
             ctx.sdcadm = self.sdcadm;
             ctx.log = self.log;
             ctx.progress = self.progress;
@@ -421,7 +424,7 @@ function do_docker(subcmd, opts, args, cb) {
                 }
                 self.progress('Creating "dockerlogger" servivce');
                 self.sdcadm.sapi.createService('dockerlogger',
-                        self.sdcadm.sdc.uuid, {
+                        self.sdcadm.sdcApp.uuid, {
                     params: {
                         image_uuid: ctx.dockerloggerImg.uuid
                     },
diff --git a/lib/post-setup/fabrics.js b/lib/post-setup/fabrics.js
index 9c58721..ab47a19 100644
--- a/lib/post-setup/fabrics.js
+++ b/lib/post-setup/fabrics.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -198,14 +198,18 @@ Fabrics.prototype.setupPortolan = function setupPortolan(cb) {
         }
     };
 
-    var app = self.sdcadm.sdc;
+    var app;
     var headnode = self.headnode;
 
     var img, haveImg, svc, svcExists, instExists;
 
     vasync.pipeline({arg: {}, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         /* @field ctx.package */
         function getPackage(ctx, next) {
+            app = self.sdcadm.sdcApp;
             var filter = {name: 'sdc_768', active: true};
             self.sdcadm.papi.list(filter, {}, function (err, pkgs) {
                 if (err) {
@@ -391,12 +395,13 @@ Fabrics.prototype.setupNat = function setupNat(cb) {
     };
 
 
-    vasync.pipeline({arg: {
-        app: self.sdcadm.sdc
-    }, funcs: [
-
+    vasync.pipeline({arg: {}, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         /* @field ctx.natPkg */
         function getNatPkg(ctx, next) {
+            ctx.app = self.sdcadm.sdcApp;
             var filter = {name: natSvcData.params.package_name, active: true};
             self.sdcadm.papi.list(filter, {}, function (err, pkgs) {
                 if (err) {
@@ -506,29 +511,39 @@ Fabrics.prototype.initDiffSchema = function initDiffSchema(cb) {
     var self = this;
     var schema = schemas.sdc.sdc_app;
     var fab, mdata;
-    var app = self.sdcadm.sdc;
-
-    if ('metadata_schemas' in app &&
-        'properties' in app.metadata_schemas &&
-        'fabric_cfg' in app.metadata_schemas.properties) {
-        mdata = app.metadata_schemas;
-        fab = mdata.properties.fabric_cfg;
-        if (jsprim.deepEqual(fab, schema.properties.fabric_cfg)) {
-            self.alreadySetup = true;
-            return cb(null);
+
+    self.sdcadm.ensureSdcApp({}, function (appErr) {
+        if (appErr) {
+            cb(appErr);
+            return;
         }
-    }
 
-    self.sdcadm.sapi.updateApplication(app.uuid, {
-        action: 'update',
-        metadata_schema: schema
-    }, function (err, sdcApp) {
-        if (err) {
-            return cb(new errors.SDCClientError(err, 'sapi'));
+        var app = self.sdcadm.sdcApp;
+
+        if ('metadata_schemas' in app &&
+            'properties' in app.metadata_schemas &&
+            'fabric_cfg' in app.metadata_schemas.properties) {
+            mdata = app.metadata_schemas;
+            fab = mdata.properties.fabric_cfg;
+            if (jsprim.deepEqual(fab, schema.properties.fabric_cfg)) {
+                self.alreadySetup = true;
+                return cb(null);
+            }
         }
-        self.sdcadm.sdc = sdcApp;
-        return cb(null);
+
+        self.sdcadm.sapi.updateApplication(app.uuid, {
+            action: 'update',
+            metadata_schema: schema
+        }, function (err, sdcApp) {
+            if (err) {
+                return cb(new errors.SDCClientError(err, 'sapi'));
+            }
+            self.sdcadm.sdcApp = sdcApp;
+            return cb(null);
+        });
     });
+
+
 };
 
 
@@ -667,32 +682,45 @@ Fabrics.prototype.checkNatPool = function checkNatPool(cb) {
 Fabrics.prototype.updateSapi = function updateSapi(cb) {
     var self = this;
 
-    if (self.sdcadm.sdc.metadata.fabric_cfg && !self.options.reconfigure) {
-        self.progress('Fabric configuration already in SAPI');
-        self.progress('Please, use \'--reconfigure\' if you want to override');
-        return cb(null);
-    }
+    self.sdcadm.ensureSdcApp({}, function (err) {
+        if (err) {
+            cb(err);
+            return;
+        }
 
-    if (self.sdcadm.sdc.metadata.fabric_cfg &&
-            jsprim.deepEqual(self.sdcadm.sdc.metadata.fabric_cfg, self.data)) {
-        self.progress('Exactly the same fabric configuration already in SAPI');
-        return cb(null);
-    }
-    self.configChanged = true;
+        if (self.sdcadm.sdcApp.metadata.fabric_cfg &&
+            !self.options.reconfigure) {
+            self.progress('Fabric configuration already in SAPI');
+            self.progress('Please, use \'--reconfigure\' ' +
+                'if you want to override');
+            return cb(null);
+        }
 
-    var word = (self.sdcadm.sdc.metadata.fabric_cfg) ? 'Updating' : 'Adding';
-    self.progress('%s fabric configuration', word);
-    /*
-     * Note, we're updating the entire application here, but update today only
-     * ever goes one layer deep. eg. update will always replace our key,
-     * 'fabric_cfg', with one that's always what we give it. In this case, it
-     * shouldn't merge anything. If that behavior changes, we're in trouble and
-     * the docs don't exactly promise one behavior or another...
-     */
-    self.sdcadm.sapi.updateApplication(self.sdcadm.sdc.uuid, {
-        action: 'update',
-        metadata: { fabric_cfg: self.data }
-    }, errors.sdcClientErrWrap(cb, 'sapi'));
+        if (self.sdcadm.sdcApp.metadata.fabric_cfg &&
+                jsprim.deepEqual(self.sdcadm.sdcApp.metadata.fabric_cfg,
+                    self.data)) {
+            self.progress('Exactly the same fabric configuration ' +
+                'already in SAPI');
+            return cb(null);
+        }
+        self.configChanged = true;
+
+        var word = (self.sdcadm.sdcApp.metadata.fabric_cfg) ? 'Updating' :
+            'Adding';
+        self.progress('%s fabric configuration', word);
+        /*
+         * Note, we're updating the entire application here, but update today
+         * only ever goes one layer deep. eg. update will always replace our
+         * key, 'fabric_cfg', with one that's always what we give it. In this
+         * case, it shouldn't merge anything. If that behavior changes, we're
+         * in trouble and the docs don't exactly promise one behavior or
+         * another...
+         */
+        self.sdcadm.sapi.updateApplication(self.sdcadm.sdcApp.uuid, {
+            action: 'update',
+            metadata: { fabric_cfg: self.data }
+        }, errors.sdcClientErrWrap(cb, 'sapi'));
+    });
 };
 
 
@@ -751,32 +779,39 @@ Fabrics.prototype.updateFabricsSvcs = function updateFabricsSvcs(cb) {
  */
 Fabrics.prototype.checkDocker = function checkDocker(cb) {
     var self = this;
-    self.sdcadm.getSvc({
-        app: self.sdcadm.sdc.uuid,
-        svc: 'docker',
-        allowNone: true
-    }, function (err, docker) {
-        if (err) {
-            return cb(err);
-        }
-        if (!docker) {
-            return cb();
+    self.sdcadm.ensureSdcApp({}, function (appErr) {
+        if (appErr) {
+            cb(appErr);
+            return;
         }
 
-        self.dockerSetup = true;
-        if (!docker.metadata.USE_FABRICS) {
-            self.sdcadm.sapi.updateService(docker.uuid, {
-                action: 'update',
-                metadata: { USE_FABRICS: true }
-            }, function (er2, _docker) {
-                if (er2) {
-                    return cb(new errors.SDCClientError(er2, 'sapi'));
-                }
+        self.sdcadm.getSvc({
+            app: self.sdcadm.sdcApp.uuid,
+            svc: 'docker',
+            allowNone: true
+        }, function (err, docker) {
+            if (err) {
+                return cb(err);
+            }
+            if (!docker) {
                 return cb();
-            });
-        } else {
-            return cb();
-        }
+            }
+
+            self.dockerSetup = true;
+            if (!docker.metadata.USE_FABRICS) {
+                self.sdcadm.sapi.updateService(docker.uuid, {
+                    action: 'update',
+                    metadata: { USE_FABRICS: true }
+                }, function (er2, _docker) {
+                    if (er2) {
+                        return cb(new errors.SDCClientError(er2, 'sapi'));
+                    }
+                    return cb();
+                });
+            } else {
+                return cb();
+            }
+        });
     });
 };
 
diff --git a/lib/post-setup/ha-binder.js b/lib/post-setup/ha-binder.js
index 8443b17..a0b51c0 100644
--- a/lib/post-setup/ha-binder.js
+++ b/lib/post-setup/ha-binder.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -86,7 +86,7 @@ function do_ha_binder(subcmd, opts, args, cb) {
 
     }
 
-    var app = self.sdcadm.sdc;
+    var app;
     var img, svc, instances, history;
     var vms;
     var oldVms;
@@ -107,7 +107,11 @@ function do_ha_binder(subcmd, opts, args, cb) {
     var hasManatee21 = false;
 
     vasync.pipeline({arg: arg, funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         function getBinderSvc(_, next) {
+            app = self.sdcadm.sdcApp;
             self.progress('Getting SDC\'s binder details from SAPI');
             self.sdcadm.sapi.listServices({
                 name: 'binder',
diff --git a/lib/post-setup/ha-manatee.js b/lib/post-setup/ha-manatee.js
index 8b7cba5..3acc58c 100644
--- a/lib/post-setup/ha-manatee.js
+++ b/lib/post-setup/ha-manatee.js
@@ -128,7 +128,7 @@ function do_ha_manatee(subcmd, opts, args, cb) {
     }
 
 
-    var app = self.sdcadm.sdc;
+    var app;
     // This is the primary instance VM:
     var pri;
     // This is the secondary instance VM, if it exists when we run the process
@@ -169,7 +169,11 @@ function do_ha_manatee(subcmd, opts, args, cb) {
     var duplicatedServers = false;
 
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.sdcadm.ensureSdcApp({}, next);
+        },
         function checkTargetServer(_, next) {
+            app = self.sdcadm.sdcApp;
             self.progress('Verifying target severs "%j" exist', opts.servers);
             self.sdcadm.cnapi.listServers(function (sErr, servers_) {
                 if (sErr) {
diff --git a/lib/post-setup/underlay-nics.js b/lib/post-setup/underlay-nics.js
index 4b319be..6e125ba 100644
--- a/lib/post-setup/underlay-nics.js
+++ b/lib/post-setup/underlay-nics.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -54,6 +54,9 @@ UnderlayNics.prototype.execute = function (opts, cb) {
     var errs = [];
 
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            sdcadm.ensureSdcApp({}, next);
+        },
         function getNapiVmImgs(_, next) {
             sdcadm.getImgsForSvcVms({
                 svc: 'napi'
@@ -186,7 +189,8 @@ UnderlayNics.prototype.execute = function (opts, cb) {
         // CNs must have the configured underlay network tag assigned to an
         // actual NIC in order to be able to add underlay-nic for the CN:
         function filterServersWithoutUnderlayNicTag(_, next) {
-            var underlayTag = sdcadm.sdc.metadata.fabric_cfg.sdc_underlay_tag;
+            var underlayTag =
+                sdcadm.sdcApp.metadata.fabric_cfg.sdc_underlay_tag;
             var cnsToSkip = [];
             var theCns = cns2Update;
             cns2Update = [];
diff --git a/lib/procedures/update-agent-v1.js b/lib/procedures/update-agent-v1.js
index 4e1b551..2cff685 100644
--- a/lib/procedures/update-agent-v1.js
+++ b/lib/procedures/update-agent-v1.js
@@ -107,6 +107,9 @@ UpdateAgentV1.prototype.execute = function uaExecute(opts, callback) {
         };
 
         vasync.pipeline({arg: context, funcs: [
+            function ensureSdcApp(_, next) {
+                sdcadm.ensureSdcApp({}, next);
+            },
             /*
              * Check if cn-agent service is already on SAPI. Otherwise,
              * inform the user how to add it.
@@ -114,7 +117,7 @@ UpdateAgentV1.prototype.execute = function uaExecute(opts, callback) {
             function checkCnAgentSvcOnSapi(_, next) {
                 sdcadm.getSvc({
                     svc: 'cn-agent',
-                    app: sdcadm.sdc.uuid,
+                    app: sdcadm.sdcApp.uuid,
                     allowNone: true
                 }, function (err, svc) {
                     if (err) {
diff --git a/lib/procedures/update-dockerlogger.js b/lib/procedures/update-dockerlogger.js
index fdd7048..8b25679 100644
--- a/lib/procedures/update-dockerlogger.js
+++ b/lib/procedures/update-dockerlogger.js
@@ -105,6 +105,9 @@ UpdateDockerlogger.prototype.execute = function udExecute(opts, callback) {
         };
 
         vasync.pipeline({arg: context, funcs: [
+            function ensureSdcApp(_, next) {
+                sdcadm.ensureSdcApp({}, next);
+            },
             /*
              * Check if docker service is already on SAPI. Otherwise,
              * inform the user how to add it.
@@ -112,7 +115,7 @@ UpdateDockerlogger.prototype.execute = function udExecute(opts, callback) {
             function checkDockerSvcOnSapi(_, next) {
                 sdcadm.getSvc({
                     svc: 'docker',
-                    app: sdcadm.sdc.uuid,
+                    app: sdcadm.sdcApp.uuid,
                     allowNone: true
                 }, function (err, service) {
                     if (err) {
@@ -141,7 +144,8 @@ UpdateDockerlogger.prototype.execute = function udExecute(opts, callback) {
                         return next();
                     }
                     progress('Creating "dockerlogger" servivce');
-                    sdcadm.sapi.createService('dockerlogger', sdcadm.sdc.uuid, {
+                    sdcadm.sapi.createService('dockerlogger',
+                        sdcadm.sdcApp.uuid, {
                         params: {
                             image_uuid: change.image.uuid
                         },
diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 2cca542..24e8cd9 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -208,19 +208,19 @@ function SdcAdm(options) {
     Object.defineProperty(this, 'updates', {
         get: function () {
             if (self._updates === undefined) {
-                assert.object(self.sdc, 'self.sdc (the SAPI "sdc" app) '
+                assert.object(self.sdcApp, 'self.sdcApp (the SAPI "sdc" app) '
                     + 'must be retrieved for client config');
                 var opts = {
                     url: self.config.updatesServerUrl,
-                    proxy: self.sdc.metadata.http_proxy || false,
+                    proxy: self.sdcApp.metadata.http_proxy || false,
                     userAgent: self.userAgent,
                     log: self.log,
                     headers: {
                         'x-request-id': self.uuid
                     }
                 };
-                if (self.sdc.metadata.update_channel) {
-                    opts.channel = self.sdc.metadata.update_channel;
+                if (self.sdcApp.metadata.update_channel) {
+                    opts.channel = self.sdcApp.metadata.update_channel;
                 }
                 self._updates = new sdcClients.IMGAPI(opts);
             }
@@ -230,11 +230,11 @@ function SdcAdm(options) {
     Object.defineProperty(this, 'imagesJo', {
         get: function () {
             if (self._imagesJo === undefined) {
-                assert.object(self.sdc, 'self.sdc (the SAPI "sdc" app) '
+                assert.object(self.sdcApp, 'self.sdcApp (the SAPI "sdc" app) '
                     + 'must be retrieved for client config');
                 var opts = {
                     url: 'https://images.joyent.com',
-                    proxy: self.sdc.metadata.http_proxy || false,
+                    proxy: self.sdcApp.metadata.http_proxy || false,
                     userAgent: self.userAgent,
                     log: self.log,
                     headers: {
@@ -343,9 +343,51 @@ function SdcAdm(options) {
     this.sadm_urError = null;
     this.sadm_urStart = null;
 }
+/*
+ * Avoid preloading sdcApp by default since we cannot load it during
+ * headnode recovery. Additionally, this allow us to handle the error
+ * messages output when the command in use is given JSON option:
+ *
+ * When `arg.quiet` option is given, this function will avoid printing
+ * any error messages to stdout and will leave the responsibility to
+ * inform the user about the exeception to the calling command.
+ */
+SdcAdm.prototype.ensureSdcApp = function (arg, cb) {
+    assert.object(arg, 'arg');
+    assert.optionalBool(arg.quiet, 'arg.quiet');
+    assert.func(cb, 'cb');
+    var self = this;
+    if (self.sdcApp) {
+        cb();
+        return;
+    }
+
+    self.getApp({app: 'sdc'}, function (appErr, app) {
+        if (appErr) {
+            // Couple known issues we can help operators with a friendly
+            // message instead of the default "ENO..." errors:
+            if (appErr.message && !arg.quiet) {
+                var msg;
+                if (appErr.message.match(/getaddrinfo ENOTFOUND/)) {
+                    msg = 'Binder service seems to be down. ' +
+                            'Please review it before proceeding';
+                } else if (appErr.message.match(/connect ECONNREFUSED/)) {
+                    msg = 'SAPI service seems to be down. ' +
+                            'Please review it before proceeding';
+                }
+                console.log(msg);
+            }
+            return cb(appErr);
+        }
+        self.sdcApp = app;
+        cb();
+    });
+
+};
+
 
 // This function defines the sdcadm properties which require async callbacks
-// to be used: 'config', 'history' and 'sdc' application.
+// to be used: 'config' and 'history'.
 SdcAdm.prototype.init = function init(cb) {
     var self = this;
     common.loadConfig({log: self.log}, function (err, config) {
@@ -359,25 +401,7 @@ SdcAdm.prototype.init = function init(cb) {
         }
 
         self.history = new History({sdcadm: self});
-
-        self.getApp({app: 'sdc'}, function (appErr, app) {
-            if (appErr) {
-                // Couple known issues we can help operators with a friendly
-                // message instead of the default "ENO..." errors:
-                if (appErr.message) {
-                    if (appErr.message.match(/getaddrinfo ENOTFOUND/)) {
-                        console.log('Binder service seems to be down. ' +
-                                'Please review it before proceeding');
-                    } else if (appErr.message.match(/connect ECONNREFUSED/)) {
-                        console.log('SAPI service seems to be down. ' +
-                                'Please review it before proceeding');
-                    }
-                }
-                return cb(appErr);
-            }
-            self.sdc = app;
-            return self.history.init(cb);
-        });
+        self.history.init(cb);
     });
 };
 
@@ -537,6 +561,9 @@ SdcAdm.prototype.listInsts = function listInsts(opts, cb) {
         insts: []
     };
     vasync.pipeline({arg: context, funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
         function getServers(ctx, next) {
             ctx.serverFromUuid = {};
             ctx.serverAdminIpFromUuid = {};
@@ -771,7 +798,7 @@ SdcAdm.prototype.listInsts = function listInsts(opts, cb) {
                 return next();
             }
             self.getSvc({
-                app: self.sdc.uuid,
+                app: self.sdcApp.uuid,
                 svc: 'dockerlogger',
                 allowNone: true
             }, function (svcErr, dlSvc) {
@@ -855,15 +882,17 @@ SdcAdm.prototype.getServices = function getServices(opts, cb) {
     assert.object(opts, 'opts');
     assert.func(cb, 'cb');
 
-    var app = self.sdc;
     var svcs = [];
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
         function getSapiSvcs(_, next) {
             // 'cloudapi' zones typically don't have `tags.smartdc_core=true`
             // so we can't filter on that. And VMAPI doesn't support filtering
             // on presence of a tag (e.g. `smartdc_role`.)
             var filters = {
-                application_uuid: app.uuid
+                application_uuid: self.sdcApp.uuid
             };
 
             if (opts.type) {
@@ -967,13 +996,32 @@ SdcAdm.prototype.getImage = function getImage(opts, cb) {
     assert.string(opts.uuid, 'opts.uuid');
     assert.func(cb, 'cb');
     var self = this;
+    var img;
 
-    self.imgapi.getImage(opts.uuid, function (iErr, iImg) {
-        if (iErr && iErr.body && iErr.body.code === 'ResourceNotFound') {
-            self.updates.getImage(opts.uuid, cb);
-        } else {
-            cb(iErr, iImg);
+    vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
+        function getImg(_, next) {
+            self.imgapi.getImage(opts.uuid, function (iErr, iImg) {
+                if (iErr && iErr.body &&
+                    iErr.body.code === 'ResourceNotFound') {
+                    self.updates.getImage(opts.uuid, function (err, image) {
+                        if (err) {
+                            next(err);
+                            return;
+                        }
+                        img = image;
+                        next();
+                    });
+                } else {
+                    img = iImg;
+                    next(iErr);
+                }
+            });
         }
+    ]}, function (pipeErr) {
+        cb(pipeErr, img);
     });
 };
 
@@ -990,8 +1038,8 @@ SdcAdm.prototype.getApp = function getApp(opts, cb) {
     assert.string(opts.app, 'opts.app');
     assert.func(cb, 'cb');
 
-    if (opts.app === 'sdc' && this.sdc) {
-        cb(null, this.sdc);
+    if (opts.app === 'sdc' && this.sdcApp) {
+        cb(null, this.sdcApp);
     } else if (common.UUID_RE.test(opts.app)) {
         this.sapi.getApplication(opts.app, errors.sdcClientErrWrap(cb, 'sapi'));
     } else {
@@ -1176,27 +1224,40 @@ SdcAdm.prototype.getImgsForSvcVms = function getImgsForSvcVms(opts, cb) {
 SdcAdm.prototype.getDefaultChannel = function getDefaultChannel(cb) {
     var self = this;
 
-    var app = self.sdc;
+    var channel;
 
-    if (self.updates.channel) {
-        cb(null, self.updates.channel);
-    } else if (app.metadata.update_channel) {
-        cb(null, app.metadata.update_channel);
-    } else {
-        self.updates.listChannels({}, function (err, channels) {
-            if (err) {
-                var e = new errors.SDCClientError(err, 'imgapi');
-                return cb(e);
-            }
+    vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
+        function getChannel(_, next) {
+            if (self.updates.channel) {
+                channel = self.updates.channel;
+                next();
+            } else if (self.sdcApp.metadata.update_channel) {
+                channel = self.sdcApp.metadata.update_channel;
+                next();
+            } else {
+                self.updates.listChannels({}, function (err, channels) {
+                    if (err) {
+                        var e = new errors.SDCClientError(err, 'imgapi');
+                        next(e);
+                        return;
+                    }
 
-            var remote = channels.filter(function (c) {
-                return (c['default']);
-            }).shift();
+                    var remote = channels.filter(function (c) {
+                        return (c['default']);
+                    }).shift();
+                    channel = remote.name;
+                    next();
+                });
+            }
 
-            return cb(null, remote.name);
-        });
+        }
+    ]}, function (pipeErr) {
+        cb(pipeErr, channel);
+    });
 
-    }
 };
 
 
@@ -1299,6 +1360,9 @@ SdcAdm.prototype.updateAgents = function updateAgents(options, callback) {
     };
 
     vasync.pipeline({arg: context, funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
         /*
          * Check for Ur availability first, as we cannot proceed without
          * it:
@@ -3246,7 +3310,7 @@ SdcAdm.prototype.genUpdatePlan = function genUpdatePlan(options, cb) {
                             if (ch.service === 'dockerlogger') {
                                 ch.service = {
                                     name: 'dockerlogger',
-                                    application_uuid: self.sdc.uuid,
+                                    application_uuid: self.sdcApp.uuid,
                                     type: 'other',
                                     params: {
                                         // First dockerlogger image ever:
@@ -4478,7 +4542,6 @@ SdcAdm.prototype.dcMaintStatus = function dcMaintStatus(cb) {
     var self = this;
     var log = self.log;
 
-    var sdcApp = self.sdc;
     var services = {};
     var maint = false;
     var cloudapiMaint;
@@ -4486,9 +4549,12 @@ SdcAdm.prototype.dcMaintStatus = function dcMaintStatus(cb) {
     var startTime;
 
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
         function getCloudapiSvc(_, next) {
             var filters = {
-                application_uuid: sdcApp.uuid,
+                application_uuid: self.sdcApp.uuid,
                 name: 'cloudapi'
             };
             self.sapi.listServices(filters, function (svcsErr, svcs) {
@@ -4504,7 +4570,7 @@ SdcAdm.prototype.dcMaintStatus = function dcMaintStatus(cb) {
 
         function getDockerSvc(_, next) {
             var filters = {
-                application_uuid: sdcApp.uuid,
+                application_uuid: self.sdcApp.uuid,
                 name: 'docker'
             };
             self.sapi.listServices(filters, function (svcsErr, svcs) {
@@ -4569,11 +4635,11 @@ SdcAdm.prototype.dcMaintStatus = function dcMaintStatus(cb) {
             if (startTime) {
                 status.startTime = startTime;
             }
-            if (sdcApp.metadata.DC_MAINT_MESSAGE) {
-                status.message = sdcApp.metadata.DC_MAINT_MESSAGE;
+            if (self.sdcApp.metadata.DC_MAINT_MESSAGE) {
+                status.message = self.sdcApp.metadata.DC_MAINT_MESSAGE;
             }
-            if (sdcApp.metadata.DC_MAINT_ETA) {
-                status.eta = sdcApp.metadata.DC_MAINT_ETA;
+            if (self.sdcApp.metadata.DC_MAINT_ETA) {
+                status.eta = self.sdcApp.metadata.DC_MAINT_ETA;
             }
             cb(null, status);
         }
@@ -4599,7 +4665,6 @@ SdcAdm.prototype.dcMaintStart = function dcMaintStart(opts, cb) {
     var self = this;
     var progress = opts.progress || function () {};
 
-    var sdcApp = self.sdc;
     var services = {};
     var headnode;
     var putCloudapiIntoMaint = false;
@@ -4607,6 +4672,9 @@ SdcAdm.prototype.dcMaintStart = function dcMaintStart(opts, cb) {
     var startTime;
 
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
         function getHeadnode(_, next) {
             self.cnapi.listServers({
                 headnode: true
@@ -4624,7 +4692,7 @@ SdcAdm.prototype.dcMaintStart = function dcMaintStart(opts, cb) {
                 return next();
             }
             var filters = {
-                application_uuid: sdcApp.uuid,
+                application_uuid: self.sdcApp.uuid,
                 name: 'cloudapi'
             };
             self.sapi.listServices(filters, function (svcsErr, svcs) {
@@ -4641,7 +4709,7 @@ SdcAdm.prototype.dcMaintStart = function dcMaintStart(opts, cb) {
                 return next();
             }
             var filters = {
-                application_uuid: sdcApp.uuid,
+                application_uuid: self.sdcApp.uuid,
                 name: 'docker'
             };
             self.sapi.listServices(filters, function (svcsErr, svcs) {
@@ -4745,11 +4813,11 @@ SdcAdm.prototype.dcMaintStart = function dcMaintStart(opts, cb) {
         function setDCMaintenance(_, next) {
             var dcMaintOpts = {};
             if (opts.message &&
-                    sdcApp.metadata.DC_MAINT_MESSAGE !== opts.message) {
+                    self.sdcApp.metadata.DC_MAINT_MESSAGE !== opts.message) {
                 dcMaintOpts.DC_MAINT_MESSAGE = opts.message;
             }
             if (opts.eta &&
-                    sdcApp.metadata.DC_MAINT_ETA !== opts.eta) {
+                    self.sdcApp.metadata.DC_MAINT_ETA !== opts.eta) {
                 dcMaintOpts.DC_MAINT_ETA = opts.eta;
             }
             if (!Object.keys(dcMaintOpts).length) {
@@ -4757,7 +4825,7 @@ SdcAdm.prototype.dcMaintStart = function dcMaintStart(opts, cb) {
             }
             progress('Saving data center maintenance eta/message');
             startTime = new Date();
-            self.sapi.updateApplication(sdcApp.uuid, {
+            self.sapi.updateApplication(self.sdcApp.uuid, {
                 metadata: dcMaintOpts
             }, function (err, app) {
                 if (err) {
@@ -4880,13 +4948,15 @@ SdcAdm.prototype.dcMaintStop = function dcMaintStop(opts, cb) {
     var progress = opts.progress || function () {};
 
     var headnode;
-    var sdcApp = self.sdc;
     var services = {};
     var disableCloudapiMaint = false;
     var disableDockerMaint = false;
     var disableAppMaint = false;
 
     vasync.pipeline({funcs: [
+        function ensureSdcApp(_, next) {
+            self.ensureSdcApp({}, next);
+        },
         function getHeadnode(_, next) {
             self.cnapi.listServers({
                 headnode: true
@@ -4901,7 +4971,7 @@ SdcAdm.prototype.dcMaintStop = function dcMaintStop(opts, cb) {
 
         function getCloudapiSvc(_, next) {
             var filters = {
-                application_uuid: sdcApp.uuid,
+                application_uuid: self.sdcApp.uuid,
                 name: 'cloudapi'
             };
             self.sapi.listServices(filters, function (svcsErr, svcs) {
@@ -4915,7 +4985,7 @@ SdcAdm.prototype.dcMaintStop = function dcMaintStop(opts, cb) {
 
         function getDockerSvc(_, next) {
             var filters = {
-                application_uuid: sdcApp.uuid,
+                application_uuid: self.sdcApp.uuid,
                 name: 'docker'
             };
             self.sapi.listServices(filters, function (svcsErr, svcs) {
@@ -4982,8 +5052,8 @@ SdcAdm.prototype.dcMaintStop = function dcMaintStop(opts, cb) {
         },
 
         function checkIfAppInMaint(_, next) {
-            if (!sdcApp.metadata.DC_MAINT_MESSAGE &&
-                !sdcApp.metadata.DC_MAINT_ETA) {
+            if (!self.sdcApp.metadata.DC_MAINT_MESSAGE &&
+                !self.sdcApp.metadata.DC_MAINT_ETA) {
                 progress('Data center is not in maintenance mode');
             } else {
                 disableAppMaint = true;
@@ -5027,7 +5097,7 @@ SdcAdm.prototype.dcMaintStop = function dcMaintStop(opts, cb) {
                 return next();
             }
             progress('Clearing data center maintenance information');
-            self.sapi.updateApplication(sdcApp.uuid, {
+            self.sapi.updateApplication(self.sdcApp.uuid, {
                 metadata: {
                     DC_MAINT_MESSAGE: null,
                     DC_MAINT_ETA: null
@@ -5104,8 +5174,6 @@ SdcAdm.prototype.dcMaintStop = function dcMaintStop(opts, cb) {
 
 SdcAdm.prototype.checkConfig = function (opts, cb) {
     var self = this;
-    // SAPI values for sdc application:
-    var sdc = self.sdc.metadata;
     // Name of SAPI services for VMs:
     var services;
     // Headnode sysinfo:
@@ -5114,6 +5182,8 @@ SdcAdm.prototype.checkConfig = function (opts, cb) {
     var admin;
     var external;
 
+    var sdc;
+
     // Errors:
     var errs = [];
 
@@ -5136,7 +5206,8 @@ SdcAdm.prototype.checkConfig = function (opts, cb) {
                     errs.push('SAPI sdc admin_nic did not match with GZ ' +
                         'Admin MAC Address');
                 }
-                if (sysinfo['Network Interfaces'][k].ip4addr !== sdc.admin_ip) {
+                if (sysinfo['Network Interfaces'][k].ip4addr !==
+                    sdc.admin_ip) {
                     errs.push('SAPI sdc admin_ip did not match with GZ ' +
                         'Admin IPv4 Address');
                 }
@@ -5315,68 +5386,69 @@ SdcAdm.prototype.checkConfig = function (opts, cb) {
 
     }
 
-    self.sapi.listServices({
-        application_uuid: sdc.uuid
-    }, function (err2, res2) {
-        if (err2) {
-            return cb(err2);
-        }
-        if (!res2.length) {
-            return cb('Cannot find SDC services in SAPI');
-        }
-
-        services = res2.filter(function (s) {
-            return (s.type === 'vm');
-        }).map(function (s) {
-            return (s.name);
+    function getVmServiceNames(_, next) {
+        self.getServices({type: 'vm'}, function (err, svcs) {
+            if (err) {
+                next(err);
+                return;
+            }
+            services = svcs.filter(function (s) {
+                return s.type === 'vm';
+            }).map(function (s) {
+                return (s.name);
+            });
+            sdc = self.sdcApp.metadata;
+            next();
         });
+    }
 
-        vasync.pipeline({
-            funcs: [
-                getSysinfo,
-                getNetworks,
-                getDcFromUfds,
-                getUfdsAdmin,
-                getVmsIps
-            ]
-        }, function (err4, _res) {
-            if (err4) {
-                return cb(err4);
-            }
 
-            // PEDRO: Note the exceptions listed below. I bet we could
-            // remove most of these variables anyway, and left a single
-            // value for *_pw.
-            services.forEach(function (s) {
-                if (!sdc[s + '_root_pw'] && s !== 'manta' && s !== 'sapi') {
-                    errs.push(sprintf('Missing %s_root_pw in SAPI', s));
-                }
+    vasync.pipeline({
+        funcs: [
+            getVmServiceNames,
+            getSysinfo,
+            getNetworks,
+            getDcFromUfds,
+            getUfdsAdmin,
+            getVmsIps
+        ]
+    }, function (err4, _res) {
+        if (err4) {
+            return cb(err4);
+        }
 
-                if (!sdc[s + '_admin_ips'] && s !== 'cloudapi' &&
-                    s !== 'manta' && s !== 'sdcsso') {
-                    errs.push(sprintf('Missing %s_admin_ips in SAPI', s));
-                }
+        // PEDRO: Note the exceptions listed below. I bet we could
+        // remove most of these variables anyway, and left a single
+        // value for *_pw.
+        services.forEach(function (s) {
+            if (!sdc[s + '_root_pw'] && s !== 'manta' && s !== 'sapi') {
+                errs.push(sprintf('Missing %s_root_pw in SAPI', s));
+            }
 
-                if (s !== 'manatee' && s !== 'binder' &&
-                    s !== 'manta' && s !== 'cloudapi') {
-                    if (!sdc[s + '_domain']) {
-                        errs.push(sprintf('Missing %s_domain in SAPI', s));
-                    }
-                    if (!sdc[s.toUpperCase() + '_SERVICE']) {
-                        errs.push(sprintf('Missing %s_SERVICE in SAPI',
-                                s.toUpperCase()));
-                    }
+            if (!sdc[s + '_admin_ips'] && s !== 'cloudapi' &&
+                s !== 'manta' && s !== 'sdcsso') {
+                errs.push(sprintf('Missing %s_admin_ips in SAPI', s));
+            }
+
+            if (s !== 'manatee' && s !== 'binder' &&
+                s !== 'manta' && s !== 'cloudapi') {
+                if (!sdc[s + '_domain']) {
+                    errs.push(sprintf('Missing %s_domain in SAPI', s));
                 }
-            });
-            // Check that ufds_remote_ip is present if this is not master:
-            if (!sdc.ufds_is_master || sdc.ufds_is_master === 'false') {
-                if (!sdc.ufds_remote_ip) {
-                    errs.push('Missing SAPI variable "ufds_remote_ip"');
+                if (!sdc[s.toUpperCase() + '_SERVICE']) {
+                    errs.push(sprintf('Missing %s_SERVICE in SAPI',
+                            s.toUpperCase()));
                 }
             }
-            return self.ufds.close(function (err3) {
-                return cb(null, errs);
-            });
+        });
+        // Check that ufds_remote_ip is present if this is not master:
+        if (!sdc.ufds_is_master || sdc.ufds_is_master === 'false') {
+            if (!sdc.ufds_remote_ip) {
+                errs.push('Missing SAPI variable "ufds_remote_ip"');
+            }
+        }
+        return self.ufds.close(function (err3) {
+            return cb(null, errs);
         });
     });
 };
@@ -5510,6 +5582,7 @@ SdcAdm.prototype.checkHealth = function checkHealth(opts, cb) {
         if (opts.type) {
             svcOpts.type = opts.type;
         }
+
         self.getServices(svcOpts, function (err, svcs) {
             if (err) {
                 if (!err.message) {  // TODO(trentm): why this?!
@@ -5753,6 +5826,13 @@ SdcAdm.prototype.checkHealth = function checkHealth(opts, cb) {
 
     vasync.pipeline({
         funcs: [
+            function ensureSdcApp(_, next) {
+                if (opts.json) {
+                    self.ensureSdcApp({ quiet: true}, next);
+                } else {
+                    self.ensureSdcApp({}, next);
+                }
+            },
             connectToUr,
             lookupServices,
             lookupInstances,
diff --git a/lib/steps/agent-services.js b/lib/steps/agent-services.js
index afd2df0..ef90e18 100644
--- a/lib/steps/agent-services.js
+++ b/lib/steps/agent-services.js
@@ -41,6 +41,7 @@ function agentServicesEnsureCreated(arg, cb) {
     assert.func(arg.progress, 'arg.progress');
     assert.object(arg.log, 'arg.log');
     assert.object(arg.sdcadm, 'arg.sdcadm');
+    assert.object(arg.sdcadm.sdcApp, 'arg.sdcaadm.sdcApp');
     assert.func(cb, 'cb');
 
     var log = arg.log.child({component: 'agentServicesEnsureCreated'}, true);
@@ -50,7 +51,7 @@ function agentServicesEnsureCreated(arg, cb) {
     // We need at least a MIN_VALID_SAPI_VERSION image so
     // type=agent suport is there.
     var MIN_VALID_SAPI_VERSION = '20140703';
-    var app = sdcadm.sdc;
+    var app = sdcadm.sdcApp;
     var historyItem = null;
 
     var img;
diff --git a/lib/steps/no-rabbit.js b/lib/steps/no-rabbit.js
index 3df7cfb..75eebb1 100644
--- a/lib/steps/no-rabbit.js
+++ b/lib/steps/no-rabbit.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -35,6 +35,7 @@ function noRabbitEnable(arg, callback) {
     assert.func(arg.progress, 'arg.progress');
     assert.object(arg.log, 'arg.log');
     assert.object(arg.sdcadm, 'arg.sdcadm');
+    assert.object(arg.sdcadm.sdcApp, 'arg.sdcadm.sdcApp');
     var log = arg.log;
     var sdcadm = arg.sdcadm;
     var progress = arg.progress;
@@ -42,8 +43,8 @@ function noRabbitEnable(arg, callback) {
     var imgsFromSvcName = {};
     var MIN_NO_RABBIT_VERSION = '20150917';
 
-    if (arg.sdcadm.sdc.metadata.no_rabbit) {
-        log.debug({no_rabbit: arg.sdcadm.sdc.metadata.no_rabbit},
+    if (arg.sdcadm.sdcApp.metadata.no_rabbit) {
+        log.debug({no_rabbit: arg.sdcadm.sdcApp.metadata.no_rabbit},
             'no_rabbit already enabled');
         return callback();
     }
@@ -128,7 +129,7 @@ function noRabbitEnable(arg, callback) {
                     no_rabbit: true
                 }
             };
-            sdcadm.sapi.updateApplication(sdcadm.sdc.uuid, update,
+            sdcadm.sapi.updateApplication(sdcadm.sdcApp.uuid, update,
                 errors.sdcClientErrWrap(next, 'sapi'));
         },
         function stopDatacenterMaint(ctx, next) {
-- 
2.21.0

